_______________________________________
# Задачи  к семинару № 2

## Задача 1: 
### 1.1 Скачать файл в уроке.  
### 1.2 Считать данные с помощью pandas.  

In [324]:
import pandas as pd

df = pd.read_csv(r'kc-house-data-kc-house-data.csv')

### 1.3 Вывести на экран первые 5 строк.  

In [325]:
df.head(5)

id             date   price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900         3       1.00         1180   
1  6414100192  20141209T000000  538000         3       2.25         2570   
2  5631500400  20150225T000000  180000         2       1.00          770   
3  2487200875  20141209T000000  604000         4       3.00         1960   
4  1954400510  20150218T000000  510000         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

 
### 1.4 Посмотреть на описание признаков и на их содержание

In [326]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  object 
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

_______________________________________________
## Задача 2:
### 2.1 Проведите первичный анализ данных:

In [327]:
df.describe()

id      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  2.161300e+04   
mean   4.580302e+09      3.370842      2.114757   2079.899736  1.510697e+04   
std    2.876566e+09      0.930062      0.770163    918.440897  4.142051e+04   
min    1.000102e+06      0.000000      0.000000    290.000000  5.200000e+02   
25%    2.123049e+09      3.000000      1.750000   1427.000000  5.040000e+03   
50%    3.904930e+09      3.000000      2.250000   1910.000000  7.618000e+03   
75%    7.308900e+09      4.000000      2.500000   2550.000000  1.068800e+04   
max    9.900000e+09     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136     84.402258  98077.939805   
std      828.090978     442.575043     29.373411    401.679240     53.505026   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1951.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4820.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000   21613.000000   21613.000000  
mean      47.560053   -122.213896    1986.552492   12768.455652  
std        0.138564      0.140828     685.391304   27304.179631  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471000   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.230000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

### 2.2 Изучите типы данных:

In [328]:
df.dtypes

id                 int64
date              object
price             object
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

### 2.3 Найдите количество пропущенных ячеек в данных:

In [329]:
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

### 2.4 Посчитайте основные статистики по всем признакам и поизучайте их:

In [330]:
df['price'] = round(df['price'].str.replace(',', '.').astype('float'), 3)

In [331]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

### 2.5 Пишите выводы:

Таблица содержит  содержит данные о недвижимости,  
таблица состоит из 21 613 строк и 21 столбеца:

|Исходные данные | Перевод| 
|----------------|--------|
|__id__  	     |         Идентификатор  |
|__date__    	 |         Дата  |
|__price__   	 |         Цена  |
|__bedrooms__    |          Спальни  |
|__bathrooms__   |          Ванные комнаты  |
|__sqft_living__ |          Жилая площадь в квадратных футах  |
|__sqft_lot__    |	      Площадь участка в квадратных футах  |
|__floors__      |          Количество этажей  |
|__waterfront__  |          Вид на воду (1 - да, 0 - нет)  |
|__view__        |          Вид (от 0 до 4)  |
|__condition__   |          Состояние (от 1 до 5)  |
|__grade__       |          Класс (от 1 до 13)  |
|__sqft_above__   |         Высота  |
|__sqft_basement__ |        Площадь подвала   |
|__yr_built__       |       Год постройки  |
|__yr_renovated__  |        Год реконструкции  |
|__zipcode__       |        Почтовый индекс  |
|__lat__           |        Широта |
|__long__          |        Длинный   |
|__sqft_living15__ |        Жилая площадь в квадратных футах  |
|__sqft_lot15__    |        Площадь участка в квадратных футах  |

Типы данных некоторых столбцов отличаются от ожидаемых  
prise -> object, ожидается float или и int  
bathrooms/floors и т.д - > float ожидается int, имеется дробная часть, что является аномалией

Не информативны имена некоторых столбцов

In [332]:
f"Минимальная стоимость: {df['price'].min()}; Максимальная стоимость: {df['price'].max()}; Средняя стоимость: {round(df['price'].mean(), 2)};"

'Минимальная стоимость: 75000.0; Максимальная стоимость: 7700000.0; Средняя стоимость: 540088.14;'

Ячейки с значением null отсутствуют.

_______________________________________________
## Задача 3:


### 3.1 В каком диапазоне изменяются стоимости недвижимости?

В диапазоне от __75000__ у.е до __7700000__ у.е

### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [333]:
living = df['sqft_living'].sum()
lot = df['sqft_lot'].sum()
result = living / lot * 100
f'{round(result, 2)}%'

'13.77%'

### 3.3 Как много домов с разными этажами в данных?

In [334]:
df['floors'].nunique()

6

В данных наблюдается __6__ вариантов этажности домов

### 3.4 Насколько хорошие состояния у домов в данных?

In [335]:
df['condition'].mean()

3.4094295100171195

В среднем состояние всех домов является удовлетворительным.

In [336]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

Преобладающие количество домов прибывают в удовлетворительном состояние.

### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [337]:
df['yr_built'].min(), df['yr_built'].max()

(1900, 2015)

Первый дом был построен в 1900 году, последний в 2015 году.

_____________________________
## Задача 4
Ответьте на несколько вопросов:


### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [338]:
round(df[df['bedrooms'] == 2]['price'].mean(), 2)

401372.68

Дома с двумя спальнями в среднем стоят: __401372.68__ у.е

### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [339]:
round(df[df['price'] > 600000]['sqft_lot'].mean(), 4)

20442.5248

Средняя площадь домов стоимостью более 600 000 у.е составляет: 20442.5248 ft²

### 4.3 Как много домов коснулся ремонт?

In [340]:
df[df['yr_renovated'] != 0].shape[0]

914

Реновация затронула __914__ домов.

### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [341]:
round(df[df['grade'] > 10]['price'].mean() - df[df['grade'] < 4]['price'].mean(), 2)

1488885.12

В среднем стоимость домов с рейтингом выше 10 отличается от домов с рейтингом ниже 4 на __1488885.12__ у.е

_____________________________
## Задача 5: 
Ответьте на несколько вопросов:

### 5.1 Выберите дом клиенту:
### Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [342]:
df[(df['waterfront'] == 1) & (df['bathrooms'] > 2) & (df['sqft_basement'] > 0)].shape[0]

76

У клиента есть __76__ вариантов домов.

### 5.2 Выберите дом клиенту:
### Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [343]:
min_price =  df[((df['view'] == 4) | (df['waterfront'] == 1)) & (df['yr_built'] > 1980) & (df['condition'] == 5)]['price'].min()
max_price =  df[((df['view'] == 4) | (df['waterfront'] == 1)) & (df['yr_built'] > 1980) & (df['condition'] == 5)]['price'].max()
min_price, max_price

(1295000.0, 3000000.0)

Цены на дома с такими параметрами лежат в диапазоне от __1295000__ у.е до __3000000__ у.е

### 5.3 Выберите дом клиенту
### Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [344]:
round(df[(df['waterfront'] < 1) & (df['floors'] == 2) & (df['price'] < 150000)]['condition'].mean(), 2)

2.83

Средняя оценка таких домов ниже средней по совокупности, близится к неудовлетворительной.
_____________________________________________________________________